### ISSUE TITLE
THC variance kernel runs slow and does not scale well with reduction dimension

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#13192]
## 🚀 Feature
Merge `torch.mean` and `torch.var` kernels on GPU.

## Motivation

`torch.var` GPU kernel launch config does not scale with larger reduction size. Meanwhile, `torch.mean` performs really well and the math are similar (`torch.var` uses welford so it is pratically the same single pass reduction).
At larger scale I am seeing 100x speed up.

## Pitch

We should figure out a reduction op API so we can support different math in our reduction op. That would enable us to merge `torch.mean` and `torch.var` kernel launch configurations.

## Alternatives

We can definitely improve `torch.var` kernel separately, since it has slightly more complicated math, we might expect it requires different heuristics than the one used in `torch.mean`. But this might not be that much of a problem. 

## Additional context

I have some monkey benchmark. Comparing `torch.mean` `torch.var` and my test kernel `welford` (which does a mean/var on contiguous dimension)
I have a 2d input tensor and does reduction on each dimension separately.

![1](https://user-images.githubusercontent.com/3709243/47595875-683c2180-d937-11e8-9663-7cf1a6736a24.png)

**Comment by jjsjann123** [comment]
I'm currently working on this. Will issue a PR once I got it into decent shape.

Here's the code to repro (with my test kernel part remove)
```
import torch

import timeit
import time
import numpy as np

import matplotlib.pyplot as plt

WIDTH = 1.0
ALPHA = 1.0

def run_func(func, inp_t, *args):
  if len(args) == 0:
    out = func(inp_t)
  elif len(args) == 1:
    out = func(inp_t, args[0])
  torch.cuda.synchronize()


class Timer:
  def __init__(self):
    self.data = np.array([])

  def new_item(self, t):
    self.data = np.append(self.data, t)

  def mean(self):
    return self.data.mean()
    
  def median(self):
    return np.percentile(self.data, 50)

def pyplot_draw(data, xticks, title, xlabel, ylabel, log, axis=None):
  plt.title(title)
  if axis != None:
    x = xticks
  else:
    x = range(len(xticks))
    plt.xticks(x, xticks)

  if log:
    plot = plt.plot
  else:
    plot = plt.loglog

  for key in data:
    plot(x, data[key], label=key, linewidth=WIDTH, alpha=ALPHA)
  plt.xlabel(xlabel)
  plt.ylabel(ylabel)

def draw_plot(result, axis = None, title = None, xlabel = None, ylabel = None, log = False):
  data = {}
  xticks = []
  for item in next(iter(result.values())):
    data[item] = []
  for key in result:
    if axis != None:
      xticks.append(key[axis])
    else:
      xticks.append(key)
    for item in data:
      data[item].append( result[key][item] )
  if title:
    pyplot_draw(data, xticks, title, xlabel, ylabel, log, axis)
  print(xticks)
  print(data)


################################################################################
# reduce timing
################################################################################
iteration = 100

reduce_dim_base = 128
output_dim_base = 64

reduce_dim_iter = 10
output_dim_iter = 5



for i in range(output_dim_iter):
  output_dim = output_dim_base * (2**i)
  configs = []
  result = {}
  
  plt.subplot(output_dim_iter, 1, i+1)

  for j in range(reduce_dim_iter):
    reduce_dim = reduce_dim_base * (2**j)
    configs.append((reduce_dim, output_dim))
  
  for reduce_dim, output_dim in configs:
    result[(reduce_dim, output_dim)] = {}
    res = result[(reduce_dim, output_dim)]
  
    torch.cuda.cudart().cudaProfilerStart()
  
    inp_t = torch.randn(output_dim, reduce_dim).requires_grad_().cuda()
    timer = Timer()
    for i in range(iteration):
      timer.new_item(timeit.timeit('run_func(torch.mean, inp_t, 1)', number=1, globals=globals()))
    res['contig_dim_mean'] = timer.median()
    timer = Timer()
    for i in range(iteration):
      timer.new_item(timeit.timeit('run_func(torch.var, inp_t, 1)', number=1, globals=globals()))
    res['contig_dim_var'] = timer.median()
  
    inp_t = torch.randn(reduce_dim, output_dim).requires_grad_().cuda()
    timer = Timer()
    for i in range(iteration):
      timer.new_item(timeit.timeit('run_func(torch.mean, inp_t, 0)', number=1, globals=globals()))
    res['noncontig_dim_mean'] = timer.median()
    timer = Timer()
    for i in range(iteration):
      timer.new_item(timeit.timeit('run_func(torch.var, inp_t, 0)', number=1, globals=globals()))
    res['noncontig_dim_var'] = timer.median()
  
    torch.cuda.cudart().cudaProfilerStop()
  
  draw_plot(result, 0, "reduction_output_size_{0}".format(output_dim), "reduction_size", "median time")
  plt.legend()
plt.show()
#========================end batchnorm timing===================================

```

**Comment by zou3519** [comment]
@jjsjann123 looking forward to it!

**Comment by zou3519** [comment]
Fixed by #13435 but please feel free to reopen if this is not the case
